crates/lib/src/qvm/libquil.rs

```diff
@@ -0,0 +1,184 @@
+//! Provides QVM functionality via libquil
+
+use std::collections::HashMap;
+use std::convert::TryFrom;
+
+use crate::RegisterData;
+
+use super::{
+    http::{self, AddressRequest},
+    QvmOptions,
+};
+
+/// The errors that can arise when using this client
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Error when calling QVM
+    #[error("error when calling QVM: {0}")]
+    LibquilSysQvm(#[from] libquil_sys::qvm::Error),
+    /// Error when calling Quilc
+    #[error("error when calling quilc: {0}")]
+    LibquilSysQuilc(#[from] libquil_sys::quilc::Error),
+    /// We currently only support requesting a specific set of register indices
+    /// or _all_ register indices.
+    #[error("can only request explicit or all indices for multishot programs")]
+    UnsupportedIndicesRequestType,
+    /// Error raised when trying to cast one integer type into another
+    #[error("could not cast value: {0}")]
+    InvalidCast(#[from] std::num::TryFromIntError),
+}
+
+impl From<Error> for super::Error {
+    fn from(error: Error) -> Self {
+        Self::Qvm {
+            message: error.to_string(),
+        }
+    }
+}
+
+/// A libquil client providing QVM functionality
+#[derive(Debug, Copy, Clone)]
+pub struct Client;
+
+#[async_trait::async_trait]
+impl crate::qvm::Client for Client {
+    async fn get_version_info(&self, _options: &QvmOptions) -> Result<String, super::Error> {
+        let version = libquil_sys::qvm::get_version_info().map_err(Error::LibquilSysQvm)?;
+        Ok(version.to_string())
+    }
+
+    async fn run(
+        &self,
+        request: &http::MultishotRequest,
+        _options: &QvmOptions,
+    ) -> Result<http::MultishotResponse, super::Error> {
+        let program = request
+            .compiled_quil
+            .parse()
+            .map_err(Error::LibquilSysQuilc)?;
+        let addresses = request
+            .addresses
+            .iter()
+            .map(|(address, indices)| match indices {
+                AddressRequest::Indices(indices) => Ok((
+                    address.clone(),
+                    libquil_sys::qvm::MultishotAddressRequest::Indices(indices.clone()),
+                )),
+                AddressRequest::IncludeAll => Ok((
+                    address.clone(),
+                    libquil_sys::qvm::MultishotAddressRequest::All,
+                )),
+                AddressRequest::ExcludeAll => Err(Error::UnsupportedIndicesRequestType),
+            })
+            .collect::<Result<_, _>>()?;
+        let result =
+            libquil_sys::qvm::multishot(&program, addresses, i32::from(request.trials.get()))
+                .map_err(Error::LibquilSysQvm)?;
+        let mut registers = HashMap::new();
+        for (address, values) in result {
```

Comment:
On mobile, so multiline suggestions are difficult, but you should be able to replace this loop with an iterator that collects into a `HashMap`. It may even be slightly more optimized since iterators provide size hints that collections can use to reserve the right capacity all at once.

This is mostly a style suggestion though, and is technically fine as-is. For the optimization part, you may want to use `HashMap::with_capacity(result.len())` above